File: typescript/tsfind/src/fileresult.ts
/*
 * fileresult.js
 *
 * encapsulates a file to be found
 */

'use strict';

import {FileType} from './filetype';

export class FileResult {
    containerSeparator = '!';
    containers: string[] = [];
    filePath: string;
    fileType: FileType;
    fileSize: number = 0;
    lastMod: number = 0;

    constructor(filePath: string, fileType: FileType, fileSize: number, lastMod: number) {
        this.filePath = filePath;
        this.fileType = fileType;
        this.fileSize = fileSize;
        this.lastMod = lastMod;
    }

    public toString(): string {
        let s = '';
        if (this.containers.length > 0) {
            s = this.containers.join(this.containerSeparator) + this.containerSeparator;
        }
        s += this.filePath;
        return s;
    }
}

File: fsharp/FsFind/FsFindLib/FileResult.fs
namespace FsFindLib

open System

module FileResult = 

    type t = {
        Containers : string list;
        FilePath : string;
        Type: FileType
        Size : int64
        LastMod : DateTime Option
    }

    let Create (filePath : string) (fileType : FileType) (size : int64) (lastMod : DateTime Option) : t =
        {
            Containers=[];
            FilePath=filePath;
            Type=fileType;
            Size=size;
            LastMod=lastMod
        }

    let ToString (sf : t) : string =
        let container_str = 
            match sf.Containers with
            | [] -> ""
            | _  -> sprintf "%s!" (String.concat "!" sf.Containers)
        $"%s{container_str}%s{sf.FilePath}"
;;
